Get file from parent and name

jcmfernandes opened the pull request:
Method `#file_from_parent_and_name` is useful because when one selects a folder using box's javascript content picker, it provides a folder object, not a path. I'm facing situations where I need to get the file object using a folder ID and a file name, hence I propose adding this method.

`#file_exists?` is more of a convenience.

xhocquet commented:
Hey @jcmfernandes, thanks for the contribution again! 😄 

I'm not so sure about this addition. My current interpretation of this library is basically a 1:1 mapping with Box's API. Following that model, it is hard to justify these kinds of methods without worrying about having to add additional features. 

For example, `file_from_parent_and_name` does not support recursion. Should it? In my applications it would need to. `file_exists?` also makes some strong assumptions about just two use-cases. Currently I believe. this belongs in the business layer with the developer's application code, not in the gem. I would strongly reconsider if Box provided these features explicitly.

jcmfernandes commented:
Hey @xhocquet! Thanks for your feedback, I really do appreciate it.

Regarding `#file_from_parent_and_name`, I feel like it's a natural complement to `#file_from_path` and `#folder_from_path`, themselves helper methods and not interfaces to specific box' endpoints.

With that said, as you pointed the need for recursivity, I agree and it made me think. So, what if, instead I add an optional parameter to `#file_from_path` and `#folder_from_path` called `starting_folder`? It will look like as follows:

```patch
From 7e6f1a2aad35dabd4bf94ebf29420bd82e66bbee Mon Sep 17 00:00:00 2001
From: Joao Fernandes <jfernandes@salsify.com>
Date: Tue, 12 Nov 2019 00:02:05 +0000
Subject: [PATCH] Allow #file_from_path and #folder_from_path to start from a
 given folder

---
 lib/boxr/files.rb   | 4 ++--
 lib/boxr/folders.rb | 4 ++--
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/lib/boxr/files.rb b/lib/boxr/files.rb
index 1023f6b..35123a6 100644
--- a/lib/boxr/files.rb
+++ b/lib/boxr/files.rb
@@ -1,7 +1,7 @@
 module Boxr
   class Client
 
-    def file_from_path(path)
+    def file_from_path(path, starting_folder = Boxr::ROOT)
       if(path.start_with?('/'))
         path = path.slice(1..-1)
       end
@@ -9,7 +9,7 @@ module Boxr
       path_items = path.split('/')
       file_name = path_items.slice!(-1)
 
-      folder = folder_from_path(path_items.join('/'))
+      folder = folder_from_path(path_items.join('/'), starting_folder)
 
       files = folder_items(folder, fields: [:id, :name]).files
       file = files.find { |f| f.name == file_name }
diff --git a/lib/boxr/folders.rb b/lib/boxr/folders.rb
index 94c7396..866a2eb 100644
--- a/lib/boxr/folders.rb
+++ b/lib/boxr/folders.rb
@@ -1,14 +1,14 @@
 module Boxr
   class Client
 
-    def folder_from_path(path)
+    def folder_from_path(path, starting_folder = Boxr::ROOT)
       if(path.start_with?('/'))
         path = path.slice(1..-1)
       end
 
       path_folders = path.split('/')
 
-      folder = path_folders.inject(Boxr::ROOT) do |parent_folder, folder_name|
+      folder = path_folders.inject(starting_folder) do |parent_folder, folder_name|
         folders = folder_items(parent_folder, fields: [:id, :name]).folders
         folder = folders.find { |f| f.name == folder_name }
         raise BoxrError.new(boxr_message: "Folder not found: '#{folder_name}'") if folder.nil?
-- 
2.21.0
```

I'm happy to drop `#file_exists?`. Let me know what you think! 😄

jcmfernandes commented:
Hey @xhocquet! Let me know what you think of my proposal once you have a little bit of time. Thank you!

xhocquet commented:
@jcmfernandes Thanks, I was going through the codebase and setting up some new views into our methods so that I could grasp how well we map 1:1 today and whether that limitation was just in my head. Like you said, there's already other methods that provide helper functionality, so I think your additions are fine in that sense.

I prefer the original from the commented updates, just by nature of adding new helper methods vs. modifying method signatures of core methods in the library. I'm a bit busy right now but I'll keep this in mind, and we can probably merge.
